streaming/cleanup.go

```diff
@@ -0,0 +1,318 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package streaming
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	dataflow "cloud.google.com/go/dataflow/apiv1beta3"
+	"cloud.google.com/go/dataflow/apiv1beta3/dataflowpb"
+	datastream "cloud.google.com/go/datastream/apiv1"
+	"cloud.google.com/go/datastream/apiv1/datastreampb"
+	dashboard "cloud.google.com/go/monitoring/dashboard/apiv1"
+	"cloud.google.com/go/monitoring/dashboard/apiv1/dashboardpb"
+	"cloud.google.com/go/pubsub"
+	"cloud.google.com/go/spanner"
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/utils"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/internal"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/logger"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/profiles"
+	"google.golang.org/api/iterator"
+)
+
+type ShardCleanupOptions struct {
+	Dataflow   bool
+	Datastream bool
+	Pubsub     bool
+	Monitoring bool
+}
+
+type JobCleanupOptions struct {
+	Dataflow   bool
+	Datastream bool
+	Pubsub     bool
+	Monitoring bool
+}
+
+func InitiateJobCleanup(ctx context.Context, jobCleanupOptions JobCleanupOptions, jobExecutionData JobExecutionData, shardExecutionDataList []ShardExecutionData, project string, instance string) {
+	//initiate shard level cleanup
+	for _, generatedResources := range shardExecutionDataList {
+		// converting to a differnet struct to prevent job level struct leakage into shard level actions
+		shardCleanupOptions := ShardCleanupOptions(jobCleanupOptions)
+		logger.Log.Info(fmt.Sprintf("Initiating cleanup for jobId: %s, dataShardId: %s\n", generatedResources.MigrationJobId, generatedResources.DataShardId))
+		err := initiateShardCleanup(ctx, shardCleanupOptions, generatedResources, project, instance)
+		if err != nil {
+			logger.Log.Debug(fmt.Sprintf("Unable to cleanup resources for jobId: %s, dataShardId: %s: %v\n", generatedResources.MigrationJobId, generatedResources.DataShardId, err))
+		} else {
+			logger.Log.Info(fmt.Sprintf("Successfully cleaned up resources for jobId: %v\n", generatedResources.MigrationJobId))
+		}
+	}
+	// initiate job level
+	if jobCleanupOptions.Monitoring {
+		var aggMonitoringResources internal.MonitoringResources
+		err := json.Unmarshal([]byte(jobExecutionData.AggMonitoringResources), &aggMonitoringResources)
+		if err != nil {
+			logger.Log.Debug("Unable to read aggregate monitoring metadata for deletion\n")
+		} else {
+			cleanupMonitoringDashboard(ctx, aggMonitoringResources, project)
+		}
+	}
+}
+
+func GetJobDetails(ctx context.Context, migrationJobId string, dataShardIds []string, project string, instance string) (JobExecutionData, []ShardExecutionData, error) {
+	dbURI := fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, "spannermigrationtool_metadata")
```

Comment:
better to define spannermigrationtool_metadata as a constant in utils/constants to avoids spell errors

Reply:
Done